Implementation and CLI-support for `--all-$KIND` flags

BenWiederhake opened the pull request:
This implements #3112.

This means all of the following commands are now possible and meaningful:
```
cargo build --all-bins
cargo build --all-tests
cargo test --all-tests
cargo test --all-bins
cargo check --all-bins --all-examples --all-tests --all-benches
```

The commits try to represent the incremental "propagation" of the new feature:
- core functionality (`cargo check --lib` passes)
- CLI suport (`cargo build` passes)
- additional tests (`cargo test` covers new functionality)

Note that `--all` is already reserved to mean "all packages of the workspace", so it can't be used to mean `--all-bins --all-examples --all-tests --all-benches`.

I intend to follow this up by some other PRs, so please do tell me where I could improve.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @brson (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

alexcrichton commented:
Awesome thanks for the PR @BenWiederhake! 

I'm currently traveling but I should be back next week, I'll try to review then.

BenWiederhake commented:
Thanks for the heads-up, and no hurry :)

alexcrichton commented on src/cargo/ops/cargo_install.rs:
```diff
@@ -402,13 +403,24 @@ fn find_duplicates(dst: &Path,
-            bins.iter().chain(examples)
-                       .filter_map(|t| check(t))
-                       .collect()
+            let all_bins: Vec<String> = bins.try_collect().unwrap_or_else(||
```
Could this use `{}` to delimit the closure?

matklad commented:
I'll second `--tests`, `--bins` etc, because we already have the `--all` flag, which controls an orthogonal feature, which packages are build. It would be prudent to add a test with combination of both features, like `cargo build --all --tests`

BenWiederhake commented:
@alexcrichton, re: stylistic nit: ah, yes, those braces belong there.  Added.

@alexcrichton, @matklad, re: rename: ehh, I personally find `--all-tests` more intuitive, especially since it disables the usually automatic filtering for feature-compatible targets.  But whatever, I changed it, it's only a string anyway.

@matklad, re: `cargo build --all --tests`: I added tests for the two scenarios in which `cargo build --all --examples` can occur.  I thought that building *examples* is slightly more interesting because building the *tests* can already be done with `cargo test --all --no-run`.

re: readability of diffs:  all my new commits can be found at the tip of my branch [`incremental`](https://github.com/BenWiederhake/cargo/commits/incremental), in case you want to see differences to the previous version of this PR.  For all other purposes, the currently selected branch `all-kinds` contains appropriately squashed stuff that is freshly rebased on top of `master`.  Have fun reading.

alexcrichton commented:
Looks great to me from a technical perspective, thanks again @BenWiederhake! I'll give others some time to weigh in and then we can see how to move forward.

alexcrichton commented:
Ok looks like we don't have too many more opinions, so I'll merge regardless :)

Thanks again for the PR @BenWiederhake!

@bors: r+

bors commented:
:pushpin: Commit 01782fe has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 01782fe639ab0802c2030fa0219597ceaa8850b9 -->

bors commented:
:hourglass: Testing commit 01782fe with merge b987ce4...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/cargo/build/1.0.1647), [status-travis](https://travis-ci.org/rust-lang/cargo/builds/225295344)
Approved by: alexcrichton
Pushing b987ce4f26a955595d168abdba951ab4bba63dd2 to master...
